lesson_22/orm/homework_22.py

```diff
@@ -0,0 +1,275 @@
+            if 'name' not in read.fieldnames or 'age' not in read.fieldnames:
+                raise KeyError("Missing 'name' or 'age' column in the file.")
+            students = [
+                db_model.Students(
+                    name=row['name'],
+                    age=row['age'],
+                ) for row in read
+            ]
+            db.session.add_all(students)
+
+        db.session.commit()
+        logging.info('Database populated successfully from CSV files!')
+    except (KeyError, FileNotFoundError) as err:
+        logging.error(f'Error with CSV file: {err}')
+    except SQLAlchemyError as err:
+        logging.error(f'Error populating database: {err}')
+
+
+def random_enrollment_assignment(db):
+    """Random enrollment assignments."""
+    try:
+        students = db.session.query(db_model.Students).all()
+        courses = db.session.query(db_model.Courses).all()
+
+        if not students or not courses:
+            logging.warning('No students or courses available for enrollment.')
+            return
+
+        unique_enrollments = set()
+        enrollments = []
+        # 3 is a random value that is used for 3 cycles of enrollments
+        for _ in range(len(students) * 3):
+            student = random.choice(students)
+            course = random.choice(courses)
+            key = (student.student_id, course.course_id)
+
+            if key not in unique_enrollments:
+                unique_enrollments.add(key)
+                enrollments.append(db_model.
+                                   Enrollments(student_id=student.student_id,
+                                               course_id=course.course_id))
+
+        db.session.add_all(enrollments)
+        db.session.commit()
+        logging.info('Random enrollment assignments completed successfully!')
+    except SQLAlchemyError as err:
+        db.session.rollback()
+        logging.error(f'Error : {err}')
+
+    finally:
+        db.session.close()
+
+
+def add_new_student(db, name, age, course_ids):
+    """Add new student and assign courses."""
+    try:
+        new_student = db_model.Students(name=name, age=age)
+        db.session.add(new_student)
+        db.session.commit()
+
+        courses = (db.session.query(db_model.Courses).
+                   filter(db_model.Courses.course_id.in_(course_ids)).all())
+        if len(courses) != len(course_ids):
+            raise ValueError(f'Some course IDs are invalid: {course_ids}')
+
+        enrollments = [
+            db_model.Enrollments(student_id=new_student.student_id,
+                                 course_id=course_id) for course_id in course_ids
+        ]
+
+        db.session.add_all(enrollments)
+        db.session.commit()
+        logging.info(f'Student {name} added.')
+    except SQLAlchemyError as err:
+        db.session.rollback()
+        logging.error(f'Error adding new student: {err}')
+
+    finally:
+        db.session.close()
```

Comment:
repetitive calls looks as something that need to be improved / re-organized.
Here possible for example to use decorator . Each function should have deco annotation. Inside decorator we do exceptons handling + finally block